File: BCW3/java/Date/Date.java
public class Date {
    private int day;
    private int month;
    private int year;
    
    private void validate(int day, int month, int year) throws InvalidDateException {
        int [] days = new int [] {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
        
        if ( year % 400 == 0 || (year % 100 != 0 && year % 4 == 0) ) {
            days[2] = 29;
        }
    
        if ( year < 1582 ) {
            throw new InvalidDateException("Invalid Year!");
        }
    
        if ( month < 1 || month > 12 ) {
            throw new InvalidDateException("Invalid Month!");
        }
    
        if ( day < 1 || day > days[month] ) {
            throw new InvalidDateException("Invalid Day!");
        }
    }
    
    public Date (int day, int month, int year) throws InvalidDateException {
        validate(day, month, year);
        
        this.day = day;
        this.month = month;
        this.year = year;
    }
    
    public Date () throws InvalidDateException {
        this(1, 1, 1970);
    }
    
    public int getDay() {
        return day;
    }
    
    public int getMonth() {
        return month;
    }
    
    public int getYear() {
        return year;
    }
    
    public String toString() {
        StringBuffer out = new StringBuffer();
        
        if ( day < 10 ) {
            out.append("0");
        }
        
        out.append(day + ".");
        
        if ( month < 10 ) {
            out.append("0");
        }
        
        out.append(month + "." + year);
        
        return out.toString();
    }
}

File: BCW3/C++/Date/Date.cpp
#include <iostream>
#include <iomanip>

#include "Date.h"

void Date::validate(int day, int month, int year) {
    int days[] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    
    if ( year % 400 == 0 || (year % 100 != 0 && year % 4 == 0) ) {
        days[2] = 29;
    }
    
    if ( year < 1582 ) {
        throw InvalidDate("Invalid Year!");
    }
    
    if ( month < 1 || month > 12 ) {
        throw InvalidDate("Invalid Month!");
    }
    
    if ( day < 1 || day > days[month] ) {
        throw InvalidDate("Invalid Day!");
    }
}

Date::Date(int day, int month, int year) {
    this->validate(day, month, year);
    
    this->day = day;
    this->month = month;
    this->year = year;
}

Date::~Date() {}

int Date::getDay() const {
    return this->day;
}

int Date::getMonth() const {
    return this->month;
}

int Date::getYear() const {
    return this->year;
}

std::ostream& operator<<(std::ostream& out, const Date& date) {
    std::cout << std::setfill('0') << std::setw(2) << date.getDay() << '.' << std::setw(2) << date.getMonth()
    << '.' << date.getYear();
    
    return out;
}
